File: AMoAdiOSSdkDemo(Objective-C)/AMoAdiOSSdkDemo(Objective-C)/ListViewTVController.m
//
//  TableViewController.m
//  AMoAdiOSSdkDemo(Object-C)
//
//  Created by AMoAd on 2016/09/12.
//  Copyright © 2016年 AMoAd. All rights reserved.
//

#import "ListViewTVController.h"
#import "AMoAdNativeView.h"

@interface ListViewTVController () <UITableViewDataSource, UITableViewDelegate, AMoAdNativeListViewDelegate>
@property int itemNo;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property NSArray* items;
@end

@implementation ListViewTVController
static NSString *kUserTableViewCell = @"UserTableViewCell";
static NSString *kAdTableViewCell = @"AdTableViewCell";
static NSString *kTag = @"ad001";
static int kBeginIndex = 2;
static int kInterval = 4;
- (void)viewDidLoad {
    [super viewDidLoad];
    [self initTableView];
    [self initAd];
    [self addItems];
}

- (void)initTableView {
    self.items = @[];
    self.items = [[NSArray alloc]init];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    [self.tableView registerNib:[UINib nibWithNibName:kUserTableViewCell bundle:nil] forCellReuseIdentifier:kUserTableViewCell];
    
    UIRefreshControl *refreshControl = [[UIRefreshControl alloc]init];
    [refreshControl addTarget:self action:@selector(onRefresh:) forControlEvents:UIControlEventValueChanged];
    [self.tableView addSubview:refreshControl];
}
- (void)initAd {
    self.itemNo = 1;
    [[AMoAdNativeViewManager sharedManager] prepareAdWithSid:self.sid defaultBeginIndex:kBeginIndex defaultInterval:kInterval iconPreloading:true];
    [[AMoAdNativeViewManager sharedManager] registerTableView:self.tableView sid:self.sid tag:kTag nibName:kAdTableViewCell
     ];
}
- (void)updateAd {
    [[AMoAdNativeViewManager sharedManager] updateAdWithSid:self.sid tag:kTag];
}

- (void)onRefresh:(UIRefreshControl *)refreshControl {
    [refreshControl beginRefreshing];
    self.itemNo = 1;
    self.items = @[];
    [self updateAd];
    [self addItems];
    [refreshControl endRefreshing];
}
- (void)addItems {
    NSArray *newItems = [self createItems];
    NSArray *mergedItems = [self.items arrayByAddingObjectsFromArray:newItems];
    self.items = [[AMoAdNativeViewManager sharedManager] arrayWithSid:self.sid tag:kTag originalArray:mergedItems];
    [self.tableView reloadData];
}

- (NSArray *)createItems {
    int start = self.itemNo;
    int end = self.itemNo + 10;
    NSMutableArray *items = [[NSMutableArray alloc]init];
    for(int i = start; i < end; i++) {
        NSString *item = @"ユーザデータ-";
        [items addObject:item];
    }
    return items;
}

- (void) addItemsIfNeeded:(NSIndexPath*)indexPath {
    if(indexPath.row >= self.items.count -1) {
        [self addItems];
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)NSIndexPath {
    return 116;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.items.count;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell;
    
    // データを表示する
    NSObject *item = self.items[indexPath.row];
    if([item isKindOfClass:AMoAdNativeViewItem.self]) {
        AMoAdNativeViewItem *adItem = (AMoAdNativeViewItem *)item;
        cell = [adItem tableView:tableView cellForRowAtIndexPath:indexPath delegate:self];
    } else {
        NSString *userItem = (NSString *)item;
        cell = [tableView dequeueReusableCellWithIdentifier:kUserTableViewCell forIndexPath:indexPath];
        UIImageView *iconView = (UIImageView *)[cell viewWithTag:10];
        UILabel *label1 = (UILabel *)[cell viewWithTag:11];
        iconView.image = [[UIImage alloc]initWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"item" ofType:@"png"]];
        label1.text = userItem;
    }
    [self addItemsIfNeeded:indexPath];
    return cell;
}

- (void)amoadNativeDidReceive:(NSString *)sid tag:(NSString *)tag view:(UIView *)view indexPath:(NSIndexPath *)indexPath state:(AMoAdResult)state {
    NSLog(@"amoadNativeDidReceive:%@ tag:%@ view:%@ indexPath:%@ state:%ld", sid, tag, view, indexPath, (long)state);
}

- (void)amoadNativeIconDidReceive:(NSString *)sid tag:(NSString *)tag view:(UIView *)view indexPath:(NSIndexPath *)indexPath state:(AMoAdResult)state {
    NSLog(@"amoadNativeIconDidReceive:%@ tag:%@ view:%@ indexPath:%@ state:%ld", sid, tag, view, indexPath, (long)state);
}

- (void)amoadNativeImageDidReceive:(NSString *)sid tag:(NSString *)tag view:(UIView *)view indexPath:(NSIndexPath *)indexPath state:(AMoAdResult)state {
    NSLog(@"amoadNativeImageDidReceive:%@ tag:%@ view:%@ indexPath:%@ state:%ld", sid, tag, view, indexPath, (long)state);
}

- (void)amoadNativeDidClick:(NSString *)sid tag:(NSString *)tag view:(UIView *)view indexPath:(NSIndexPath *)indexPath {
    NSLog(@"amoadNativeDidClick:%@ tag:%@ view:%@ indexPath:%@", sid, tag, view, indexPath);
}
@end

File: AMoAdiOSSdkDemo/AMoAdiOSSdkDemo/ListViewTVController.swift
//
//  ListViewTVController.swift
//  AMoAdiOSSdkDemo
//
//  Created by AMoAd on 2016/08/17.
//  Copyright © 2016年 AMoAd. All rights reserved.
//

import UIKit

class ListViewTVController: UIViewController, UITableViewDataSource, UITableViewDelegate, AMoAdNativeListViewDelegate {
    let userItemCell = "UserTableViewCell"
    let adItemCell = "AdTableViewCell"
    let beginIndex = 2
    let interval = 4
    let tag = "ad01"
    var items:NSArray = []
    var itemNo = 1
    var sid:String!
    @IBOutlet weak var tableView: UITableView!

    override func viewDidLoad() {
        super.viewDidLoad()
        initTableView()
        initAd()
        addItems()
    }

    func initTableView(){
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(UINib(nibName: userItemCell, bundle: nil), forCellReuseIdentifier: userItemCell)
        
        
        let refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: #selector(onRefresh(_:)), for: UIControlEvents.valueChanged)
        tableView.addSubview(refreshControl)
    }
    
    func initAd(){       
        AMoAdNativeViewManager.shared().prepareAd(withSid: sid, defaultBegin: beginIndex, defaultInterval: interval, iconPreloading: true)
        AMoAdNativeViewManager.shared().register(tableView, sid: sid, tag: tag, nibName: adItemCell)
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell:UITableViewCell!
        let item = items[(indexPath as NSIndexPath).row]
        if (item as AnyObject).isKind(of: AMoAdNativeViewItem.self) {
            //広告アイテム
            let adItem = item as! AMoAdNativeViewItem
            cell = adItem.tableView(tableView, cellForRowAt: indexPath, delegate: self)
        } else {
            //ユーザーアイテム
            let userItem = items[(indexPath as NSIndexPath).row]
            cell = tableView.dequeueReusableCell(withIdentifier: userItemCell, for: indexPath)
            let image = cell.viewWithTag(10) as? UIImageView
            image?.image = UIImage(named: "item")
            let title = cell.viewWithTag(11) as? UILabel
            title?.text = userItem as? String
            
        }
        addItemsIfNeeded(indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        //コンテンツの高さ
        return CGFloat(116)
    }
    
    func updateAd(){
        AMoAdNativeViewManager.shared().updateAd(withSid: sid, tag: tag)
    }
    
    func onRefresh(_ refreshControl:UIRefreshControl) {
        refreshControl.beginRefreshing()
        itemNo = 1
        items = []
        updateAd()
        addItems()
        refreshControl.endRefreshing()
    }
    
    func addItemsIfNeeded(_ indexPath:IndexPath) {
        if (indexPath as NSIndexPath).row >= items.count - 1 {
            addItems()
        }
    }
    
    func createItems() -> [String] {
        let start = itemNo
        let end = itemNo + 10
        var items = [String]()
        for n in start..<end {
            items.append("ユーザデータ\(n)")
        }
        itemNo = end
        return items
    }

    func addItems(){
        let newItems = createItems()
        let mergedItems = items.addingObjects(from: newItems)
        items = AMoAdNativeViewManager.shared().array(withSid: sid, tag: tag, originalArray: mergedItems) as NSArray
        tableView.reloadData()
    }

    func amoadNativeDidReceive(_ sid: String!, tag: String!, view: UIView!, indexPath: IndexPath!, state: AMoAdResult) {
        print("amoadNativeDidReceive(\(sid), \(tag), \(indexPath.row), \(state))")
    }
    func amoadNativeIconDidReceive(_ sid: String!, tag: String!, view: UIView!, indexPath: IndexPath!, state: AMoAdResult) {
        print("amoadNativeIconDidReceive(\(sid), \(tag), \(indexPath.row), \(state))")
    }
    func amoadNativeImageDidReceive(_ sid: String!, tag: String!, view: UIView!, indexPath: IndexPath!, state: AMoAdResult) {
        print("amoadNativeImageDidReceive(\(sid), \(tag), \(indexPath.row), \(state))")
    }
    func amoadNativeDidClick(_ sid: String!, tag: String!, view: UIView!, indexPath: IndexPath!) {
        print("amoadNativeDidClick(\(sid), \(tag), \(indexPath.row))")
    }
}
